RxJava2: Add support for changeset observables

cmelchior opened the pull request:
Part of #4991 
Fixes #4277 

This PR adds support for our fine-grained collection notifications by adding `asChangesetObservable()` methods to the relevant classes.

The name was discussed in #4277, but if anyone has better suggestions I'm all ears.

The biggest difference is that these are `Observables` compared to `Flowables` for our standard listeners. The reason for this is that our various changeset classes depend on the state of the previous one, so just throwing these away would not be acceptable. It would be up to end users to decide what to do if they cannot keep up. For animations, developers would most likely just kill the animation and jump to the new end state, but that is not up to us to decide.

TODO
- [x] Redirect to #4991  once #4992 is merged.

bmeike commented on realm/realm-library/src/main/java/io/realm/RealmResults.java:
```diff
@@ -305,6 +306,41 @@ public void removeChangeListener(OrderedRealmCollectionChangeListener<RealmResul
 
     /**
+     * Returns an Rx Observable that monitors changes to this RealmResults. It will emit the current RealmResults when
+     * subscribed to. For each update to the RealmResult a pair consisting of the RealmResults and the
```
picky, but you could leave the "to" off the end of this sentence, if you wanted.

bmeike commented on realm/realm-library/src/main/java/io/realm/RealmResults.java:
```diff
@@ -305,6 +306,41 @@ public void removeChangeListener(OrderedRealmCollectionChangeListener<RealmResul
+     * {@link OrderedCollectionChangeSet} will be sent. The changeset will be {@code null} the first
+     * time an RealmResults is emitted.
+     * <p>
+     * Changeset observables do not support backpressure as each {@link OrderedCollectionChangeSet} depends on
```
Is this comment necessary?  Isn't that the distinction between `Flowable` and `Observable`?

bmeike commented on realm/realm-library/src/main/java/io/realm/RealmResults.java:
```diff
@@ -305,6 +306,41 @@ public void removeChangeListener(OrderedRealmCollectionChangeListener<RealmResul
+     * time an RealmResults is emitted.
+     * <p>
+     * Changeset observables do not support backpressure as each {@link OrderedCollectionChangeSet} depends on
+     * the previous one. Handling backpressure is therefor up to callers of this method.
```
therefore

bmeike commented on realm/realm-library/src/main/java/io/realm/RealmResults.java:
```diff
@@ -305,6 +306,41 @@ public void removeChangeListener(OrderedRealmCollectionChangeListener<RealmResul
+    public Observable<Pair<RealmResults<E>, OrderedCollectionChangeSet>> asChangesetObservable() {
+        if (realm instanceof Realm) {
+            return realm.configuration.getRxFactory().changesetsFrom((Realm) realm, this);
+        } else if (realm instanceof DynamicRealm) {
```
The `else` seems redundant?

bmeike commented on realm/realm-library/src/main/java/io/realm/RealmResults.java:
```diff
@@ -305,6 +306,41 @@ public void removeChangeListener(OrderedRealmCollectionChangeListener<RealmResul
+            @SuppressWarnings("UnnecessaryLocalVariable")
+            Observable results = realm.configuration.getRxFactory().changesetsFrom(dynamicRealm, dynamicResults);
+            return results;
+        } else {
```
here too...

bmeike replied:
You want that var for debugging?

Zhuinden commented:
`Add two new wrapper classes for changes observables. Move Pair back t…`

Elegant!

cmelchior commented:
Updated. After discussion with @bmeike we came to the conclusion that having specifically named wrapper types for Rx emissions would increase readability, so I introduced `CollectionChange` and `ObjectChange` and moved `Pair` back into the private API.

zaki50 commented on realm/realm-library/src/main/java/io/realm/RealmList.java:
```diff
@@ -890,6 +892,36 @@ public String toString() {
+     * <p>
+     * RealmList will continually be emitted as the RealmList is updated - {@code onComplete} will never be called.
+     * <p>
+ *   * Note that when the {@link Realm} is accessed from threads other than where it was created,
```
extra `*` here

zaki50 commented on realm/realm-library/src/main/java/io/realm/internal/util/Pair.java:
```diff
@@ -34,9 +34,9 @@
 
     /**
-     * Constructor for a Pair.
+     * Constructor for a CollectionChange.
```
This should be `Pair`

zaki50 commented on realm/realm-library/src/main/java/io/realm/rx/CollectionChange.java:
```diff
@@ -0,0 +1,101 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
No need to add our license comment as well?

zaki50 commented on realm/realm-library/src/main/java/io/realm/rx/ObjectChange.java:
```diff
@@ -0,0 +1,107 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
```
No need to add our license comment as well?

dalinaum commented:
Are `ObjectChange` and `CollectionChange` from AOSP?

cmelchior commented:
No, I made them

dalinaum commented on realm/realm-library/src/main/java/io/realm/internal/util/Pair.java:
```diff
@@ -34,9 +34,9 @@
+     * Constructor for a CollectionChange.
      *
-     * @param first the first object in the Pair.
+     * @param first the first object in the CollectionChange.
```
Should `CollectionChange`s in `Pair.java` be `Pair`?

Zhuinden replied:
This should definitely be Pair

Zhuinden commented on realm/realm-library/src/androidTest/java/io/realm/RxJavaTests.java:
```diff
@@ -770,6 +770,7 @@ public void run() throws Exception {
         subscription = observable.subscribe(new Consumer<Realm>() {
             @Override
             public void accept(Realm ignored) throws Exception {
+                assertEquals(2, Realm.getLocalInstanceCount(ignored.getConfiguration()));
```
it's not ignored anymore

dalinaum commented:
@cmelchior Their license headers are for Android Open Source Project. They should be fixed.

Zhuinden commented on realm/realm-library/src/main/java/io/realm/rx/CollectionChange.java:
```diff
@@ -0,0 +1,101 @@
+ * This is used by {@link RealmResults#asChangesetObservable()}} and {@link RealmList#asChangesetObservable()} as
+ * RxJava is only capable of emitting one item, not multiple.
+ */
+public class CollectionChange<E extends OrderedRealmCollection> {
```
Is `OrderedRealmCollection` raw type on purpose?

cmelchior replied:
Not really, but it doesn't matter. When you autocomplete the `new Consumer(...)` it will use the full generic argument.
